### PR TITLE
Safeguard local_log_sum_exp optimization against -inf values

### LINKED ISSUE #461: LogSumExp of `-inf` returns `nan` instead of `-inf`
```python
x = at.vector('x')
res1 = at.log(at.sum(at.exp(x)))  # nan with sum
res2 = at.log(at.prod(at.exp(x)))  # but not with prod

fun = aesara.function([x], [res1, res2])
print(fun(np.array([-np.inf, -np.inf])))  # [array(nan), array(-inf)]
```

for reference in numpy we get `-inf`, together with a divide by zero encountered in log

```python
print(np.log(np.sum(np.exp([-np.inf, -np.inf]))))  # -inf
```

This showed up in the pymc3 `logsumexp` function which returns `nan` with this input:

```python
from pymc3.math import logsumexp

x = at.vector('x')
res = logsumexp(x)
fun = aesara.function([x], res) 

print(fun(np.array([-np.inf, -np.inf])))  # [nan]
```
Whereas the scipy reference works fine
```python
print(scipy.special.logsumexp([-np.inf, -np.inf]))  # -inf
```

Weirdly it happens with addition, but not subtraction or multiplication, so maybe the problem is not the sum but the addition?

```python
x = at.vector('x')
res1 = at.log(at.exp(x[0]) + at.exp(x[1]))
res2 = at.log(at.exp(x[0]) - at.exp(x[1]))
fun = aesara.function([x], [res1, res2])
print(fun(np.array([-np.inf, -np.inf])))  # [array(nan), array(-inf)]

### PR DESCRIPTION
**Pull request by ricardoV94**
Fixes #461

**Thank you for opening a PR!**

Here are a few important guidelines and requirements to check before your PR can be merged:
+ [x] There is an informative high-level description of the changes.
+ [x] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [x] [`pre-commit`](https://pre-commit.com/#installation) is installed and [set up](https://pre-commit.com/#3-install-the-git-hook-scripts).
+ [x] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [x] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), and there are **no commits that fix changes introduced by other commits in the same branch/BR**.  If your commit description starts with "Fix...", then you're probably making this mistake.
+ [x] There are tests covering the changes introduced in the PR.

Don't worry, your PR doesn't need to be in perfect order to submit it.  As development progresses and/or reviewers request changes, you can always [rewrite the history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) of your feature/PR branches.

If your PR is an ongoing effort and you would like to involve us in the process, simply make it a [draft PR](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests).

### CODE REVIEW

**Comment by ricardoV94** [comment]
The following expression `log(exp(x) + exp(y))` still returns `nan` `-inf` inputs. This one relies on the more limited `local_log_add` optimization which only works for the addition of two exponent terms. I tried to tinker with it in two way, both not very successfully:
1. Canonicalize to the expression `log(sum(exp([x, y])))` which would then be  stabilized by the same `log_sum_exp` in this PR, as well as to work beyond two inputs only as now. However this raises some errors in the tests.
2. Use the same stabilization logic as in the `log_sum_exp` for the two input case only. The issue is that the optimization is then applied recursively without a stop. This would require the manual `ignore_newtrees` command.

Before trying any other changes I wanted to ask for some feedback on whether it's worth the trouble, and how it should be done.

File: aesara/tensor/math_opt.py
Comment (brandonwillard) [inline aesara/tensor/math_opt.py:2287]: We don't want to remove these optimizations from the "stabilization" and "specialization" databases; otherwise, we&mdash;for instance&mdash;remove the ability to apply all stabilizations/specializations with a single optimizer query, and people are forced to either use all the `fast_run` optimizations or know about this particular optimization and apply it manually.

File: aesara/tensor/math_opt.py
Comment (brandonwillard) [inline aesara/tensor/math_opt.py:2301]: The result was previously returned as a `logp1`.  That change is what prevents the fixed-point (aka "equilibrium") optimizers (e.g. the optimizer behind "stabilize") from converging, since it produces yet another `log(exp(...) + exp(...))` form into the graph, and that matches the implicit "pattern" of `local_log_add_exp`, etc., etc.

We can rewrite this so that it doesn't return the same form it replaces, or we can attempt to determine whether or not this `node` is already in the desired form (i.e. add a guard).  The former is generally easier and more consistent.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aesara/tensor/math_opt.py
Comment (brandonwillard) [inline aesara/tensor/math_opt.py:2301]: Consider something like this:
```suggestion
                max_infs = isinf(max_pre)
                exp_sum_pre = switch(max_infs, 0, exp(add(*[p - max_pre for p in pre_exp])))
                ret = max_pre + log1p(exp_sum_pre)
```

Comment (ricardoV94) [reply aesara/tensor/math_opt.py:2301]: Thanks for the suggestion, I think there is an inherent limitation to the `log1p` trick, that does not allow for the switch safeguard nor even for more than two inputs, but I have to double check.

Comment (brandonwillard) [reply aesara/tensor/math_opt.py:2301]: See [here](https://github.com/pymc-devs/aesara/pull/465#discussion_r646232449).

Comment (ricardoV94) [reply aesara/tensor/math_opt.py:2301]: For instance if you have three inputs of `0` being exp added the log1p trick fails already

Comment (brandonwillard) [reply aesara/tensor/math_opt.py:2301]: Can you move this `switch` into the `add` you were using?  That would break the `log(add(exp(...), ...))` pattern by producing `log(add(switch(..., 0, exp(...)), ...))`.

Comment (ricardoV94) [reply aesara/tensor/math_opt.py:2301]: Maybe!

Comment (ricardoV94) [reply aesara/tensor/math_opt.py:2301]: Maybe. Will check it.

Comment (ricardoV94) [reply aesara/tensor/math_opt.py:2301]: This works, even though it feels a bit less elegant:
```python
max_pre = reduce(maximum, pre_exp)
ret = max_pre + log(add(*[
  switch(isinf(max_pre), exp(max_pre), exp(p - max_pre))
  for p in pre_exp
]))
```
Compared to the non-stable
```python
max_pre = reduce(maximum, pre_exp)
max_pre = switch(isinf(max_pre), 0, max_pre)
ret = max_pre + log(add(*[exp(p - max_pre) for p in pre_exp]))
```

What do you think?

**Comment by ricardoV94** [comment]
I pushed changes to both `local_log_add_exp` and `local_log_sum_exp` that make the opt stable by placing the switch just before the `exp`. They are as separate commits just for ease of comparison during review, but I can merge them if we decide to keep the later.

**Comment by ricardoV94** [comment]
Alternatively, would it be better to check if the graph looks like it's already under the optimization and just ignore in that case (I think that was you other suggestion @brandonwillard, right)?

**Comment by brandonwillard** [comment]
> Alternatively, would it be better to check if the graph looks like it's already under the optimization and just ignore in that case (I think that was you other suggestion @brandonwillard, right)?

No, I don't think it would necessarily be better.

**Comment by brandonwillard** [comment]
> I pushed changes to both `local_log_add_exp` and `local_log_sum_exp` that make the opt stable by placing the switch just before the `exp`. They appear as separate commits from the previous two just for ease of comparison during review, but I can merge them if we decide to keep them.

Since the latest commits undo edits from the commits before them, it would be better to merge them into single commits.

**Comment by ricardoV94** [comment]
Cleaned the commits

**Review by brandonwillard — DISMISSED** [review]
These are really useful additions, and it's great to finally have them inplace.  Much appreciated!

**Review by twiecki — APPROVED** [review]

**Comment by ricardoV94** [comment]
@brandonwillard I am still unsure about the best procedure for PRs when the main branch gets ahead of it.

If I do a local rebase, older reviews are automatically dismissed when I force-push. If I do a merge commit, they remain in place. Is there a specific procedure I can do to both rebase and not dismiss older reviews? Or should we request a review after every rebase / merge anyway.

**Comment by brandonwillard** [comment]
> @brandonwillard I am still unsure about the best procedure for PRs when the main branch gets ahead of it.
> 
> If I do a local rebase, older reviews are automatically dismissed when I force-push. If I do a merge commit, they remain in place. Is there a specific procedure I can do to both rebase and not dismiss older reviews? Or should we request a review after every rebase / merge anyway.

Yeah, request a review after rebasing.  Sometimes rebases introduce new problems, so it's often worth another look.

**Pull request merged**
